src/js_shared_storage.cpp

```
@@ -513,3 +514,78 @@ napi_value JsSharedStorage::tryToLock(napi_env env, napi_callback_info info)
 }
+
+napi_status JsSharedStorage::throw_error(napi_env env, unsigned int status,
+                                         const std::string* identifier)
+{
+    napi_status result = napi_ok;
+    std::string message;
+    std::string decoratedIdentifier;
+
+    if (identifier != nullptr)
+    {
+        decoratedIdentifier = " \"" + *identifier + "\"";
+    }
+
+    switch (status)
+    {
+    case storage::eOk:
+        break;
+
+    case storage::eCannotCreateStorage:
+        message = "cannot create the storage" + decoratedIdentifier + ". It may already exist.";
+        break;
+
+    case storage::eCannotOpenStorage:
+        message = "cannot open the storage" + decoratedIdentifier + ". It may not exist.";
+        break;
+
+    case storage::eCannotConstructItem:
+        message = "cannot set the item" + decoratedIdentifier + ". The storage may be full.";
+        break;
+
+    case storage::eCannotReplaceItem:
+        message = "cannot set the item" + decoratedIdentifier +
+                  ". An item with the same key exists and cannot be removed.";
+        break;
+
+    case storage::eCannotRemoveItem:
+        message = "cannot remove the item" + decoratedIdentifier + ".";
+        break;
+
+    case storage::eCannotClearStorage:
+        message = "cannot remove all items in the storage.";
+        break;
+
+    default:
+        result = napi_throw_error(env, nullptr, "internal storage error.");
+        break;
+    }
+
+    if (!message.empty())
+    {
+        napi_value errorMsg = nullptr;
+        napi_value errorCode = nullptr;
+        napi_value error = nullptr;
+
+        result = napi_helpers::createValueStringUTF8(message, env, &errorMsg);
+        if (result == napi_ok)
+        {
+            char buffer[16];
+            if (snprintf(buffer, 16, "%d", status) > 0)
+            {
+                result = napi_helpers::createValueStringUTF8(std::string(buffer), env, &errorCode);
+            }
+        }
+        if (result == napi_ok)
```

You may nest this if statement and the one following it at line **584** in the if of the line **572**.
```c++
if(result)
{
    result = op1();
    if(result)
    {
         result = op2();
         if(result)
         {
                ....
         }
    }
}
```

---

well, I think it's better to throw the JS error even if converting the error code to string fails. In all cases, the error will be thrown at least with a valid error message.